src/history.coffee

```diff
@@ -24,60 +26,73 @@ class History
   groupChangesSinceCheckpoint: (checkpointId) ->
     checkpointIndex = @getCheckpointIndex(checkpointId)
     return false unless checkpointIndex?
+    hasSeenChanges = false
     for entry, i in @undoStack by -1
       break if i is checkpointIndex
       if @undoStack[i] instanceof Checkpoint
-        @undoStack.splice(i, 1)
+        @undoStack.splice(i, 1) if hasSeenChanges
```

Comment:
There could be an arbitrary number of checkpoints at the top of the stack, all of which we probably want to preserve. So conceptually, what we're looking for are checkpoints that occur between changes. It's true that we could probably always make this loop one iteration shorter by starting at the penultimate entry, but I think we'd still need to keep some piece of state like `hasSeenChanges`.